wgsl/index.bs

```diff
@@ -6983,11 +6981,283 @@ TODO: *Stub*: Expression evaluation
+
+Like for statements, if the set of behaviors for an expression is a singleton, then the node corresponding to control flow at the exit of it is replaced by *CF* (the control flow at the beginning of it), overriding the rules below.
+
+<table class='data'>
+  <caption>Uniformity rules for expressions (in normal rvalue position)</caption>
+  <thead>
+    <tr><th>Expression<th>New nodes<th>Recursive analyses<th>Resulting control flow node, value node<th>New edges
+  </thead>
+  <tr><td class="nowrap">e1 || e2
+      <td>
+      <td class="nowrap">*CF*, *e1* -> *CF1*, *V1*<br>
+          *V1*, *e2* -> *CF2*, *V2*
+      <td>*CF2*, *V2*
+      <td>
+  <tr><td class="nowrap">e1 && e2
+      <td>
+      <td class="nowrap">*CF*, *e1* -> *CF1*, *V1*<br>
+          *V1*, *e2* -> *CF2*, *V2*, *B2*
+      <td>*CF2*, *V2*
+      <td>
+  <tr><td class="nowrap">Literal
+      <td>
+      <td>
+      <td>*CF*, *CF*
+      <td>
+   <tr><td class="nowrap">reference to local variable "x"
+      <td>*Result*
+      <td class="nowrap">*X* is the node corresponding to "x"
+      <td>*CF*, *Result*
+      <td class="nowrap">*Result* -> {*CF*, *X*}
+   <tr><td class="nowrap">reference to global variable "x"
+      <td>
+      <td>
+      <td>*CF*, *CannotBeUniform*
+      <td>
+   <tr><td class="nowrap">e1.field
+      <td>
+      <td class="nowrap">*CF*, *e1* -> *CF1*, *V1*
+      <td>*CF1*, *V1*
+      <td>
+   <tr><td class="nowrap">e1[e2]
+      <td> *Result*
+      <td class="nowrap">*CF*, *e1* -> *CF1*, *V1*<br>
+          *CF1*, *e2* -> *CF2*, *V2*
+      <td>*CF2*, *Result*
+      <td class="nowrap">*Result* -> {*V1*, *V2*}
+</table>
+
+<table class='data'>
+  <caption>Uniformity rules for expressions in lvalue positions</caption>
+  <thead>
+    <tr><th>Expression<th>New nodes<th>Recursive analyses<th>Resulting control flow node, variable node<th>New edges
+  </thead>
+  <tr><td class="nowrap">reference to local variable "x"
+      <td>
+      <td class="nowrap">*X* is the node corresponding to "x"
+      <td>*CF*, *X*
+      <td>
+  <tr><td class="nowrap">reference to global variable "x"
+      <td>
+      <td>
+      <td>*CF*, *CannotBeUniform*
+      <td>
+  <tr><td class="nowrap">e1.field
+      <td>
+      <td class="nowrap">*CF*, *e1* => *CF1*, *L1*
+      <td>*CF1*, *L1*
+      <td>
+  <tr><td class="nowrap">e1[e2]
+      <td>
+      <td class="nowrap">*CF*, *e1* => *CF1*, *L1*<br>
+          *CF1*, *e2* -> *CF2*, *V2*
+      <td>*CF2*, *L1*
+      <td class="nowrap">*L1* -> *V2*
+</table>
+
+### Annotating the uniformity of every point in the control-flow ### {#uniformity-optional-diagnosis-mode}
+
+This entire subsection is non-normative.
+
+If implementers want to provide developers with a diagnostic mode that shows for each point in the control-flow of the entire shader whether it is uniform or not (and thus whether it would be valid to call a function that requires uniformity there), we suggest the following:
```

Comment:
Adding some kind of annotation where developers can request special uniformity for the CF_start of a function call, or for an arg_i, would allow developers to build some kinds of static assets:

```
[[uniform]] assertInUniformCF() {
}
assertUniformValue(_: [[uniform]] T) {
}
```

Reply:
I agree, this would be a nice thing to add in a future PR.

Reply:
I agree this shouldn't block this PR.